Refactor pipeline management

ranj063 opened the pull request:
This PR refactors the pipeline management to do the following:
1. Make sure that BE pipeline widgets/routes are set up and freed during BE DAI ops. A BE pipeline is the one that contains the DAI widget
2. Pipeline trigger is already split between FE and BE DAI ops except for pipeline reset. This PR also modifies pipeline set/up free to the respective DAI ops i.e BE pipeline in the BE DAI hw_free and the rest of the pipelines in the FE DAI hw_free.
3. The widget's ipc_prepare op is intentionally left to the handled during the FE DAI hw_params because it requires propagating pipeline_params all the way from the source to the sink widget.

So the overall flow for a normal start/stop would be:

1. Pipeline widgets are prepared during FE DAI hw_params. This includes all widgets belonging to all pipelines including the BE pipeline as well.
2. All widgets and routes belonging to BE pipelines are set up during BE DAI prepare.
3. All widgets and routes belonging to the remaining pipelines are set up during FE DAI prepare.
4. All widgets and routes belonging to BE pipelines are freed and unprepared during BE DAI hw_free
5. All widgets and routes belonging to the remaining pipelines are freed and unprepared during FE DAI hw_free

The above flow ensures that the pipelines are unprepared only during hw_free. Since the hw_params do not change because of xruns/suspend, there is no need to unprepare and re-prepare the widgets in these cases. The only critical change would be to make sure that the host DMA and the link DMA are reconfigured before restarting.

In the case of an xrun, the change would be that the BE pipeline would be freed, the link DMA will be reconfigured and the pipelines will be set up again in the prepare following the xrun stop trigger. And the remaining pipelines would be freed, host DMA reconfigured and the pipelines will be set up again in the prepare ioctl following the xrun stop trigger.

The case of a suspend/resume during audio is slightly different from the xrun case: The BE pipeline will be freed during the suspend trigger. The link DMA will be reconfigured and the BE pipeline will be set up again during the BE DAI prepare after resuming. Similarly, for the remaining pipelines, they will be freed during suspend and the host DMA will be reconfigured and pipelines set up during the prepare after resume.

ranj063 committed:
ASoC: SOF: pcm: Split up widget prepare and setup

In preparation for refacting pipeline management, split the widget
prepare and set up between the hw_params and prepare ioctls. This is
required to ensure that the BE pipeline widgets can be set up during the
BE DAI prepare and the remaining widgets will be set up during the FE
DAI prepare. The widget's ipc_prepare op for all widgets in both the
BE and FE pipelines are handled during the FE DAI hw_params to make sure
that the pipeline params can be propagated all the way from the source
widget to the sink widget.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: sof-audio: Add helper functions for BE managed pipelines

Define 3 new helper functions for setting up/freeing/unpreparing BE
pipelines. These will be used when the pipeline management is
refactored to set up/free widgets, routes and pipelines during BE DAI
prepare or hw_free.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: Refactor pipeline management

Invoke the BE pipeline helper functions in the BE DAI prepare and hw_free
to ensure that all the widgets and routes belonging to the BE pipeline are
handled in the BE DAI ops. Add a new field in struct snd_sof_pipeline to
identify BE managed pipelines i.e. those that will managed by the BE DAI
ops. Lastly, modify the FE ops to make sure that the widgets/routes
belonging to these pipelines are skipped during setup/prepare/unprepare.
Note that the widget_ipc_prepare op handles all pipelines including
those managed by the BE DAI ops. This is because in order to prepare all
the widgets, the hw_params needs to be propagated along the list of
connected DAPM widgets in order from the source to the sink.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: ipc4-topology: Remove the skip_during_fe_trigger flag

Now that all widgets/routes/pipelines are setup, triggered and freed in
the BE DAI ops, we can reuse the be_pipeline flag in struct
snd_sof_pipeline to skip triggering the BE pipelines in the PCM trigger.
So remove the skip_during_fe_trigger flag from struct sof_ipc4_pipeline.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 commented:
SOFCI TEST

ranj063 committed:
drivers: soundwire: Add new prepare op in struct sdw_intel_ops

Add and define the prepare op for Intel ACE2.x platforms. For the
moment, the prepare_stream op is exactly the same as the params_stream
op. But it will be modified in the following patch when the pipeline
management is refactored.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

bardliao commented on sound/soc/sof/pcm.c:
```diff
@@ -258,6 +256,26 @@ static int sof_pcm_prepare(struct snd_soc_component *component,
+	list = spcm->stream[dir].list;
+	params = &spcm->params[substream->stream];
+	platform_params = &spcm->platform_params[substream->stream];
+	ret = sof_widget_list_setup(sdev, spcm, params, platform_params, dir);
```
As setup and prepare are split, why don't we do setup in `sof_pcm_hw_params`?

ranj063 replied:
we dont do setup in hw_params @bardliao. See my earlier comment. Its widget_prepare in hw_params and widget setup in prepare

bardliao commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -322,9 +337,20 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 	int stream = substream->stream;
+	int ret;
+
+	ret = hda_dai_hw_params(substream, &rtd->dpcm[stream].hw_params, dai);
```
Is there any specific reason that we always call hw_params in .prepare?

ranj063 replied:
because sometimes we dont have a hw_free like in the case of xruns and we'll need to do everything again.

lgirdwood replied:
prepare() should be fast. We should call this out in comments - we need to do X,Y out of cycle because Z

bardliao commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -286,6 +286,34 @@ static int intel_params_stream(struct sdw_intel *sdw,
+		return res->ops->prepare_stream(res->dev, &params_data);
+
+	return -EIO;
+}
```
Don't we need to do the same in drivers/soundwire/intel.c for MTL and older platforms?
BTW, I think it is better to split the commit into 2 commits. I.e. Add the prepare_stream callback first the then use it.

ranj063 replied:
No, we dont. MTL and older platforms do not usethe HD-Audio DMA

ranj063 replied:
yes, I do need to split up this commit and reorder it to go before the actual refactoring. I will fix it in the next iteration

bardliao reviewed (commented):
Can we call setup in hw_params and prepare in prepare?

ranj063 commented:
> Can we call setup in hw_params and prepare in prepare?

No. A widget needs to be prepared before it can be set up. So ipc_prepare needs to be done in hw_params before we do set up in prepare

ranj063 commented:
SOFCI TEST

ranj063 committed:
ASoC: SOF: sof-audio: Add a new op in struct sof_ipc_tplg_ops

Add a new host_config op in struct sof_ipc_tplg_ops and define it for
IPC4. This will be used to configure the host widget during prepare
after a suspend/resume or after an xrun.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 commented:
SOFCI TEST

bardliao commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -434,6 +434,248 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+	int ret;
+
+	/* set up the widget if it belongs to the same pipeline as the DAI widget */
+	if (swidget->spipe != spipe || !swidget->prepared)
```
Do we need to go through the widgets in the path like hda_dai_unprepare_widgets_in_pipeline() does in the `!swidget->prepared` case? And same question for hda_dai_free_widgets_in_pipeline().

ranj063 replied:
@bardliao no we dont. We're only dealing with the BE pipeline here. And once we know the DAI widget, we can find all widgets in the BE pipeline with the pipeline ID of the DAI widget

bardliao commented:
SOFCI TEST

ujfalusi commented:
@ranj063, this sequence still fails (HDA machine):
terminal 1:
`aplay -Dhw:0,0`
CTRL+z

terminal 2:
`aplay -Dhw:0,3`

terminal 1:
`fg`

```
[11038.698037] snd_soc_core:dpcm_be_dai_trigger:  Analog Playback and Capture: ASoC: trigger BE Analog Playback and Capture cmd 1
[11038.698057] snd_sof_intel_hda_common:hda_dai_trigger: sof-audio-pci-intel-tgl 0000:00:1f.3: cmd=1 dai Analog CPU DAI direction 0
[11038.698071] snd_sof:sof_ipc4_set_pipeline_state: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline instance 0 state 3
[11038.698078] snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13000003|0x0: GLB_SET_PIPELINE_STATE
[11038.698434] snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000000|0x0: GLB_SET_PIPELINE_STATE
[11038.698455] snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x13000003|0x0: GLB_SET_PIPELINE_STATE
[11038.698460] snd_sof:sof_ipc4_set_pipeline_state: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline instance 0 state 4
[11038.698464] snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13000004|0x0: GLB_SET_PIPELINE_STATE
[11038.703027] snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000007|0x0: GLB_SET_PIPELINE_STATE
[11038.703031] sof-audio-pci-intel-tgl 0000:00:1f.3: FW reported error: 7 - Unsupported operation requested
[11038.703058] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc error for msg 0x13000004|0x0
[11038.703062] sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at soc_dai_trigger on Analog CPU DAI: -22
[11038.703067]  HDA Analog: ASoC: trigger FE cmd: 1 failed: -22
```

lgirdwood reviewed (commented):
@ranj063 we you able to see any IP programming delta with debug in the hw_read(), hw_write() ?

lgirdwood commented:
@ujfalusi can you review whilst the HDA issue is being resolved. Thanks !

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1797,6 +1797,24 @@ sof_ipc4_prepare_dai_copier(struct snd_sof_dev *sdev, struct snd_sof_dai *dai,
+	struct snd_sof_widget *pipe_widget = swidget->spipe->pipe_widget;
+	struct sof_ipc4_copier_data *copier_data = &ipc4_copier->data;
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+	u32 host_dma_id = platform_params->stream_tag - 1;
```
host_dma_id can be moved under the ChainDMA case?

ranj063 replied:
its used in both cases now.

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1797,6 +1797,24 @@ sof_ipc4_prepare_dai_copier(struct snd_sof_dev *sdev, struct snd_sof_dai *dai,
+	u32 host_dma_id = platform_params->stream_tag - 1;
+
+	if (pipeline->use_chain_dma) {
+		pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_HOST_ID(host_dma_id);
```
it would be better to clear the HOST_ID bits before or-ing

ranj063 replied:
ack

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1797,6 +1797,24 @@ sof_ipc4_prepare_dai_copier(struct snd_sof_dev *sdev, struct snd_sof_dai *dai,
+{
+	struct sof_ipc4_copier *ipc4_copier = (struct sof_ipc4_copier *)swidget->private;
+	struct snd_sof_widget *pipe_widget = swidget->spipe->pipe_widget;
+	struct sof_ipc4_copier_data *copier_data = &ipc4_copier->data;
```
Do we need NULL check here?

This is going to be skipped in DSPless mode, right?

ranj063 replied:
the op will only be invoked when dspless mode is not selected. So should be OK.

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -230,6 +231,8 @@ struct sof_ipc_tplg_ops {
 	int (*dai_config)(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
 			  unsigned int flags, struct snd_sof_dai_config_data *data);
+	void (*host_config)(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
+			    struct snd_sof_platform_stream_params *platform_params);
```
Are we confident that the host_config can never fail, not even in the future?

ranj063 replied:
host config only updates the ipc structure. There's no reason for it to fail unless there's a IPC sent which happens later when we set up the host component

ujfalusi commented:
@ranj063, with a quick test this appears to work fine (no regression), but needs to be rebased because of the spcm print PR and also the this:
```diff
diff --git a/sound/soc/sof/intel/hda-dai.c b/sound/soc/sof/intel/hda-dai.c
index 915a1109b2e3..c7625b73d713 100644
--- a/sound/soc/sof/intel/hda-dai.c
+++ b/sound/soc/sof/intel/hda-dai.c
@@ -745,7 +745,7 @@ int sdw_hda_dai_prepare(struct snd_pcm_substream *substream, struct snd_pcm_hw_p
 
 	return ret;
 }
-EXPORT_SYMBOL_NS(sdw_hda_dai_prepare, SND_SOC_SOF_INTEL_HDA_COMMON);
+EXPORT_SYMBOL_NS(sdw_hda_dai_prepare, "SND_SOC_SOF_INTEL_HDA_COMMON");
 
 static int hda_dai_suspend(struct hdac_bus *bus)
 {
```

ujfalusi commented:
Suspend while pause still have problems, but it is the same without the PR.

lgirdwood commented:
> Suspend while pause still have problems, but it is the same without the PR.

We should not be pausing anymore when topology is being used to indicate pause support or not.

ujfalusi commented:
> > Suspend while pause still have problems, but it is the same without the PR.
> 
> We should not be pausing anymore when topology is being used to indicate pause support or not.

True, but if the topology indicates that then we should handle it.

ujfalusi commented on sound/soc/sof/pcm.c:
```diff
@@ -106,9 +106,9 @@ sof_pcm_setup_connected_widgets(struct snd_sof_dev *sdev, struct snd_soc_pcm_run
 		spcm->stream[dir].list = list;
 
-		ret = sof_widget_list_setup(sdev, spcm, params, platform_params, dir);
+		ret = sof_widget_list_prepare(sdev, spcm, params, platform_params, dir);
```
it is really confusing that in `setup` we `prepare` and later in `prepare` we `setup`

I'm not sure which has precedence... Do we need to prepare things to be able to setup or we need to setup to be able to prepare them.

ranj063 replied:
@ujfalusi there's a bit of confusion wrt naming between the pcm ioctl and widgets. But what we do is widget_prepare during hw_params and widget_setup during prepare because prepare succeeds hw_params

ujfalusi replied:
Right, but we will sof_widget_list_prepare()  in sof_pcm_setup_connected_widgets()
iow, we **prepare** from **setup**
and
sof_widget_list_setup() in sof_pcm_prepare()
iow we **setup** from **prepare**

I know it is all fine if you know the background on what these are, but for an untrained eye this is confusing.

ranj063 replied:
@ujfalusi I'm going to rename the widget ops in a follow up so that we dont confuse widget_prepare with PCM hw_params. But for now, yes this is confusing

lgirdwood commented:
> > > Suspend while pause still have problems, but it is the same without the PR.
> > 
> > 
> > We should not be pausing anymore when topology is being used to indicate pause support or not.
> 
> True, but if the topology indicates that then we should handle it.

There should be no pipelines/topologies supporting pause today.

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -434,6 +434,248 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+	int ret;
+
+	/* set up the widget if it belongs to the same pipeline as the DAI widget */
+	if (swidget->spipe != spipe || !swidget->prepared)
```
is this correct? you will return 0 if it is not belonging to the same pipe OR if it is _not_ prepared? What does prepared flag means here?
if it is prepared and we run setup on it it is OK to enter the function?

ranj063 replied:
@ujfalusi the prepared flag check is necessary for the aggregated amp case in SDW. For all other cases, all widgets will be prepared anyway. But in the case of aggregated amps, we have 2 DAI widgets but we only need to handle the DAI widget which is in the list of connected widgets and is prepared.

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -434,6 +434,248 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+		return;
+
+	/* skip if the widget is in use or if it is already unprepared */
+	if (!swidget || !swidget->prepared || swidget->use_count > 0)
```
at line 494 you already did a NULL dereference.

ranj063 replied:
good catch. fixed

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -434,6 +434,248 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+	int err;
+
+	/* free the widget if it belongs to the same pipeline as the DAI widget */
+	if (swidget->spipe != spipe || !swidget->prepared)
```
Here the check makes sense.

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -446,6 +688,9 @@ static const struct hda_dai_widget_dma_ops hda_ipc4_dma_ops = {
 	.get_hlink = hda_get_hlink,
+	.set_up_be_pipeline = hda_ipc4_set_up_be_pipeline,
+	.free_be_pipeline = hda_ipc4_free_be_pipeline,
+	.unprepare_be_pipeline = hda_dai_unprepare_widgets_in_pipeline,
```
So, if the BE is not serviced by HD-DMA (but with dwdma or something else) and not IPC4 I assume then we don't need to do this black magic?

ranj063 replied:
thats right because in that case, we dont have anything to do in the BE DAI ops at all.

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -434,6 +434,248 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+	return 0;
+}
+
+static int hda_ipc4_set_up_be_pipeline(struct snd_soc_dapm_widget *w, int dir)
```
there is not much IPC4 specific code in these and what they call are not IPC4 specific anymore.
I'm not sure if I follow it.

ranj063 replied:
@ujfalusi there's isnt anything IPC4 specific here but I dont want to modify anything for IPC3 anymore. So we just let IPC3 work as before without moving the pipeline management to the BE DAI ops.

ujfalusi commented on sound/soc/sof/sof-audio.c:
```diff
@@ -291,6 +291,7 @@ int sof_route_setup(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *wsourc
 	sroute->setup = true;
 	return 0;
 }
+EXPORT_SYMBOL(sof_route_setup);
```
You should have done this earlier, the previous patch uses the function.

ranj063 replied:
rebased now

ujfalusi commented on sound/soc/sof/sof-audio.c:
```diff
@@ -415,8 +438,12 @@ sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widg
+	 * if it belongs to a BE pipeline.
+	 */
+	if (!swidget || !swidget->prepared || swidget->use_count > 0 ||
+	    swidget->spipe->be_pipeline)
```
In some places use use sof_is_be_pipeline_widget() in some swidget->spipe->be_pipeline

ranj063 replied:
yes, changed to use the utility function here

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -136,9 +136,8 @@ sof_ipc4_add_pipeline_to_trigger_list(struct snd_sof_dev *sdev, int state,
-	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
 
-	if (pipeline->skip_during_fe_trigger)
+	if (spipe->be_pipeline)
```
I guess sof_is_be_pipeline_widget(pipe_widget) would be too much indirection to take, right?

ranj063 replied:
yeah, this is definitely simpler and besides that util is a static function in sof-audio.c

ujfalusi reviewed (commented):
@ranj063, I get the overall idea now.
What I'm a bit puzzled is that all vendors will need to implement the BE pipe management (which is mostly generic pipe/widget mangling code) on their own.

I'm not sure if we can do better? Some how have the BE code also up in the core and call in to platform if needed?

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -286,6 +286,27 @@ static int intel_params_stream(struct sdw_intel *sdw,
 	return -EIO;
 }
 
+static int intel_prepare_stream(struct sdw_intel *sdw,
```
Commit message:
For the moment, the **prepare_stream op** is exactly the same as the **prepare_stream op**.
I think one of them should be different, in this form it sound obvious that A is exactly the same as A.

ujfalusi replied:
also, this applies to ACE2+ platforms, I guess.

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -696,6 +798,13 @@ static int hda_dai_suspend(struct hdac_bus *bus)
+
+			if (ops->free_be_pipeline) {
+				ret = ops->free_be_pipeline(w,
+							    hdac_stream(hext_stream)->direction);
```
`hdac_stream(hext_stream)` == `s`, no?

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -336,9 +370,33 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
+
+	/* if this is a prepare without a hw_free or a suspend, free the BE pipeline */
+	if (swidget->spipe->complete && ops->free_be_pipeline) {
+		ret = ops->free_be_pipeline(w, substream->stream);
```
you have substream->stream saved as local variable.

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -336,9 +370,33 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
+	if (ops && ops->set_up_be_pipeline)
+		return ops->set_up_be_pipeline(w, substream->stream);
+
+	return ret;
```
you can return 0 from here to be more explicit (ret must be 0 if you reached this point)

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -460,10 +518,28 @@ static int non_hda_dai_hw_params(struct snd_pcm_substream *substream,
+		return ops->set_up_be_pipeline(w, stream);
 
-	return non_hda_dai_hw_params(substream, &rtd->dpcm[stream].hw_params, cpu_dai);
+	return ret;
```
Here also, just return 0

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -638,6 +714,39 @@ int sdw_hda_dai_trigger(struct snd_pcm_substream *substream, int cmd,
+	if (ops->set_up_be_pipeline)
+		return ops->set_up_be_pipeline(w, stream);
+
+	return ret;
```
return 0;

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -434,6 +422,251 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+	int ret;
+
+	/* set the be_pipeline flag true for the pipeline */
+	swidget->spipe->be_pipeline = true;
```
This will be done each time a stream starts?

ranj063 replied:
changed this to set it during ops selection

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -434,6 +422,251 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+		dev_err(sdev->dev, "failed to set up routes in the BE pipeline with DAI: %s\n",
+			w->name);
+
+	swidget->spipe->complete = 1;
```
even if it fails?

ranj063 replied:
fixed

ujfalusi commented on sound/soc/sof/sof-audio.c:
```diff
@@ -291,6 +291,18 @@ int sof_route_setup(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *wsourc
+{
+	struct snd_sof_widget *swidget = w->dobj.private;
+
+	if (swidget->spipe->be_pipeline)
```
I would rather use `be_managed_pipeline`. This flag is not set for example for IPC3 BE pipelines which does not make them non BE pipelines.

ranj063 replied:
done

ujfalusi reviewed (commented):
@ranj063, finally I think I managed to get the idea and I really like it!

However I think there could be a better and probably more intuitive way to achieve this.

I would add the hda_ipc4_set_up_be_pipeline()/hda_ipc4_free_be_pipeline()/hda_dai_unprepare_widgets_in_pipeline() and their helpers to generic SOF code (to sof-audio.c perhaps)
as snd_sof_set_up_be_pipeline()/snd_sof_free_be_pipeline()/snd_sof_unprepare_widgets_in_pipeline() and export them.

In Intel/HDA code instead of callbacks I would have a flag to not that these platforms will manage the BE pipeline themselves (IPC4)

If this flag is set then you would set the swidget->spipe->be_managed_pipeline which will cause core to skip the BE pipe and in hda code you would either use this swidget->spipe-be_managed_pipeline flag or the one which is local to intel code and call the appropriate generic helper to do the pipeline handling.

I do think that this would make it more intuitive at the end while keeping the core idea intact.

ranj063 commented:
SOFCI TEST

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -416,10 +416,11 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
-		/*
-		 * STOP/SUSPEND trigger is invoked only once when all users of this pipeline have
-		 * been stopped. So, clear the started_count so that the pipeline can be reset
-		 */
```
This comment still stands as we clear the started_count, should we keep it?

ranj063 replied:
@ujfalusi the comment was useful before because we set the started_count to 0 before the pipeline was reset in the FE DAI hw_free. But now the pipeline has just been reset right here in this op, so it is only natural to clear the started_count.

ujfalusi reviewed (approved):
@ranj063, looks much cleaner, thank you for considering my suggestions!
I have one nitpick for a removed comment
